src/bookshelf/src/bksw_sql.erl

```diff
@@ -0,0 +1,253 @@
+            Buckets;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+list_bucket(BucketName) ->
+    case sqerl:select(list_bucket_with_hash, [BucketName], rows_as_records, ?DB_FILE_TX_FM) of
+        {ok, Buckets} ->
+            Buckets;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+create_file_link_data(BucketName, Name, DataId) ->
+    case sqerl:select(create_file_link_data, [BucketName, Name, DataId], first_as_scalar, [success]) of
+        {ok, FileId} ->
+            {ok, FileId};
+        {conflict, Reason} ->
+            {error, Reason};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+link_file_data(FileId, DataId) ->
+    case sqerl:select(link_file_data, [FileId, DataId], first_as_scalar, [success]) of
+        {ok, _} ->
+            ok;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+find_file(Bucket, Name) ->
+    case sqerl:select(find_file, [Bucket, Name], {first_as_record, ?DB_FILE_TX_FM }) of
+        %% Awkward sanity check that we got back the expected record type here.
+        {ok, Object} when db_file =:= element(1, Object) ->
+            {ok, Object};
+        {ok, none} ->
+            {ok, not_found};
+        {error, Error} ->
+            {error, Error}
+    end.
+
+delete_file(FileId) ->
+    case sqerl:statement(delete_file, [FileId], count) of
+        {ok, 1} ->
+            ok;
+        Error ->
+            Error
+    end.
+
+update_metadata(DataId, Size, Chunks, SumMD5, SumSha256, SumSha512) ->
+    case sqerl:statement(update_metadata, [DataId, Size, Chunks, SumMD5, SumSha256, SumSha512], count) of
+        {ok, 1} ->
+            ok;
+        Error ->
+            Error
+    end.
+
+insert_file_data() ->
+    case sqerl:statement(insert_file_data, [], first_as_scalar, [data_id]) of
+        {ok, 1, DataId} ->
+            {ok, DataId};
+        {error, _} = Error ->
+            Error
+    end.
+
+-spec add_file_chunk(integer(), integer(), binary()) -> ok.
+add_file_chunk(DataId, Sequence, Chunk) ->
+    case sqerl:statement(add_file_chunk, [DataId, Sequence, Chunk], count) of
+        {ok, 1} ->
+            ok;
+        {conflict, _} ->
+            error("Never should happen");
+        Error ->
+            Error
+    end.
+
+get_chunk_data(Id, ChunkSequence) ->
+    case sqerl:select(fetch_file_chunk_data, [Id, ChunkSequence], first_as_scalar, [data]) of
+        {ok, none} ->
+            {ok, not_found};
+        {ok, Data} ->
+            {ok, Data};
+        {error, Error} ->
+            {error, Error}
+    end.
+
+init_transfer_state() ->
+    #file_transfer_state{
+       size = 0,
+       next_chunk = 0,
+       hash_context_md5 = crypto:hash_init(md5),
+       hash_context_sha256 = crypto:hash_init(sha256),
+       hash_context_sha512 = crypto:hash_init(sha512)
+      }.
+
+update_transfer_state(#file_transfer_state{next_chunk = NextChunk} = UploadState, Data, ChunkIncr) when size(Data) == 0 ->
```

Comment:
We should comment on this function that it's used for both streaming uploads and downloads, and perhaps change the variable name `UploadState` to something less confusing.